sdk/core/System.ServiceModel.Rest/src/Message/HttpPipelineRequest.cs

```diff
@@ -65,50 +76,50 @@ private HttpMethod ToHttpMethod(string method)
         }; ;
     }
 
-    #region Construction for transport
-
-    internal HttpRequestMessage BuildRequestMessage(PipelineMessage? message = default)
+    internal HttpRequestMessage BuildRequestMessage(CancellationToken cancellationToken)
     {
         // TODO: this is confusing where we are passing in message and also
         // using private members on the request.
 
         HttpMethod method = ToHttpMethod(Method);
-        Uri uri = message!.Request.Uri;
+        Uri uri = Uri;
         HttpRequestMessage httpRequest = new HttpRequestMessage(method, uri);
-        CancellationToken cancellationToken = message?.CancellationToken ?? default;
 
         PipelineContentAdapter? httpContent = _content != null ? new PipelineContentAdapter(_content, cancellationToken) : null;
         httpRequest.Content = httpContent;
 #if NETSTANDARD
         httpRequest.Headers.ExpectContinue = false;
 #endif
 
-        Headers.TryGetHeaders(out IEnumerable<MessageHeader<string, object>> headers);
-        foreach (MessageHeader<string, object> header in headers)
+        // TODO: Come back and address this implementation per switching on string/list
+        // header values once we reimplement assignment of headers to ResponseHeader directly.
+        Headers.TryGetHeaders(out IEnumerable<KeyValuePair<string, IEnumerable<string>>> headers);
+        foreach (KeyValuePair<string, IEnumerable<string>> header in headers)
         {
-            switch (header.Value)
+            object headerValue = header.Value.Count() == 1 ? header.Value.First() : header.Value;
```

Comment:
Using Linq here is going to hit performance hard.  Come back and address it.

Reply:
This can be addressed when we make the HttpPipelineRequest a wrapper over HttpRequestMessage instead of delaying the instantiation to the transport.